skbio/diversity/beta/_unifrac.py

```diff
@@ -189,44 +199,304 @@ def weighted_unifrac(u_counts, v_counts, otu_ids, tree, normalized=False,
+    # _vectorize_counts_and_tree takes
+    u_counts = np.asarray(u_counts)
+    v_counts = np.asarray(v_counts)
+    counts = np.vstack([u_counts, v_counts])
+    counts_by_node, tree_index, branch_lengths = \
+        _vectorize_counts_and_tree(counts, otu_ids, tree)
+    # unpack counts vectors for single pairwise UniFrac calculation
+    u_node_counts = counts_by_node[0]
+    v_node_counts = counts_by_node[1]
+
+    u_total_count = u_counts.sum()
+    v_total_count = v_counts.sum()
+
+    return (u_node_counts, v_node_counts, u_total_count, v_total_count,
+            tree_index)
+
+
+def _unweighted_unifrac(u_node_counts, v_node_counts, branch_lengths):
+    """
+    Parameters
+    ----------
+    u_node_counts, v_node_counts : np.array
+        Vectors indicating presense (value greater than zero) and absense
+        (value equal to zero) of nodes in two samples, `u` and `v`. Order is
+        assumed to be the same as in `branch_lengths`.
+    branch_lengths : np.array
+        Vector of branch lengths of all nodes (tips and internal nodes) in
+        postorder representation of their tree.
+
+    Returns
+    -------
+    float
+        Unweighted UniFrac distance between samples.
+
+    Notes
+    -----
+    The count vectors passed here correspond to all nodes in the tree, not
+    just the tips.
+
+    """
+    unique_nodes = np.logical_xor(u_node_counts, v_node_counts)
+    observed_nodes = np.logical_or(u_node_counts, v_node_counts)
+    unique_branch_length = (branch_lengths * unique_nodes).sum()
+    observed_branch_length = (branch_lengths * observed_nodes).sum()
+    if observed_branch_length == 0.0:
+        # handle special case to avoid division by zero
+        return 0.0
+    return unique_branch_length / observed_branch_length
+
+
+def _weighted_unifrac(u_node_counts, v_node_counts, u_total_count,
+                      v_total_count, branch_lengths):
+    """
+    Parameters
+    ----------
+    u_node_counts, v_node_counts : np.array
+        Vectors indicating presense (value greater than zero) and absense
+        (value equal to zero) of nodes in two samples, `u` and `v`. Order is
+        assumed to be the same as in `branch_lengths`.
+    u_total_count, v_total_counts : int
+        The sum of ``u_node_counts`` and ``v_node_counts`` vectors,
+        respectively. This could be computed internally, but since this is a
+        private method and the calling function has already generated these
+        values, this saves an iteration over each of these vectors.
+    branch_lengths : np.array
+        Vector of branch lengths of all nodes (tips and internal nodes) in
+        postorder representation of their tree.
+
+    Returns
+    -------
+    float
+        Weighted UniFrac distance between samples.
+
+    Notes
+    -----
+    The count vectors passed here correspond to all nodes in the tree, not
+    just the tips.
+
+    """
+    if u_total_count > 0:
```

Comment:
can be condensed some, or a ternary if the variable names were reduced. Same comment on `v_total_count` below

```python
u_node_proportions = u_node_counts
if u_total_count:
    u_node_proportions /= u_total_count
```